pkg/readiness/object_tracker.go

```diff
@@ -123,10 +140,32 @@ func (t *objectTracker) CancelExpect(o runtime.Object) {
 		return
 	}
 
-	delete(t.expect, k)
-	delete(t.seen, k)
-	delete(t.satisfied, k)
-	t.cancelled[k] = struct{}{}
+	t.cancelExpectNoLock(k)
+}
+
+func (t *objectTracker) TryCancelExpect(o runtime.Object) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	// Respect circuit-breaker.
+	if t.allSatisfied {
+		return
+	}
+
+	k, err := objKeyFromObject(o)
+	if err != nil {
+		log.Error(err, "skipping")
+		return
+	}
+
+	// Check if it's time to delete an expectation or just decrement its allotted retries
+	obj := t.expect[k]
+	shouldDel := obj.decrementRetries()
+	t.expect[k] = obj // set the changed obj back to the map, as the value is not a pointer
```

Comment:
What happens if we call `TryExpect` for the same object `retries+2` times? I think it will panic.

Reply:
Would it be better to add explicit conditional logic around this, as opposed to relying on the zero-value behavior?  Something like:

```
// Check if it's time to delete an expectation or just decrement its allotted retries
obj, exists := t.expect[k]
if !exists {
  log.Info("Tried to cancel non-existent expectation with key ", k.String())
  return
}

shouldDel := obj.decrementRetries()
t.expect[k] = obj // set the changed obj back to the map, as the value is not a pointer

if shouldDel {
    t.cancelExpectNoLock(k)
}

```

Reply:
Another option would be to just use pointers in the map to avoid the extra code in the first place.

If you add the if-statement, no need for the log.Info() IMO.

I like the if-statement. It seems less brittle as it wont break if `decrementRetries()` changes.

Reply:
We have to be careful - if you skip calling `cancelExpectNoLock` then the object is not marked in `t.cancelled` - which changes the old behavior for certain out-of-order scenarios.

A comment and a test would work for me, otherwise we need to define the correct behavior for cancelling unexpected objects in light of the new code.

Reply:
The out-of-order execution is a good point... I had forgotten about that. Given that, I agree the comment is the better choice.

Reply:
Wait... out-of-order also adds a new layer of difficulty here.

If we perform a `TryCancel` on something that has no expectations yet, we should probably track the # of attempts. We shouldn't short-circuit the retry behavior.

Reply:
So @maxsmythe and I have been discussing the right way to implement this and came up with the idea of tracking `TryCancelExpect` calls in their own separate map.  Our thinking was this would allow us to track the number of attempts made to cancel, regardless of the other actions that have been called on an object.

It turns out that this will provide a fix for a nasty edge case.  Because of this section in `TryCancelExpect()`:

```
obj := t.expect[k]
shouldDel := obj.decrementRetries()
t.expect[k] = obj // set the changed obj back to the map, as the value is not a pointer
```

An as yet unexpected obj will actually be _added_ to `t.expect` as a _zero-value version of the struct_.  This over-writes the retry behavior, which is set via a constructor function.

This provides further indication that separating these ideas into two maps will make the code easier to grok, as I obviously did this by mistake.  Yay for TDD!